langchain/callbacks/promptlayer_callback.py

```diff
@@ -0,0 +1,125 @@
+import datetime
+from typing import Any, Dict, List, Optional
+
+from promptlayer.utils import get_api_key, promptlayer_api_request
+
+from langchain.callbacks.base import BaseCallbackHandler
+from langchain.chat_models import ChatOpenAI
+from langchain.llms import OpenAI, Replicate
+from uuid import UUID
+from langchain.schema import (
+    AIMessage,
+    BaseMessage,
+    HumanMessage,
+    LLMResult,
+    SystemMessage,
+)
+
+
+class PromptLayerHandler(BaseCallbackHandler):
+    def _convert_message_to_dict(self, message):
+        if isinstance(message, HumanMessage):
+            message_dict = {"role": "user", "content": message.content}
+        elif isinstance(message, AIMessage):
+            message_dict = {"role": "assistant", "content": message.content}
+        elif isinstance(message, SystemMessage):
+            message_dict = {"role": "system", "content": message.content}
+        else:
+            raise ValueError(f"Got unknown type {message}")
+        if "name" in message.additional_kwargs:
+            message_dict["name"] = message.additional_kwargs["name"]
+        return message_dict
+
+    def _create_message_dicts(self, messages):
+        params: Dict[str, Any] = {}
+        message_dicts = [self._convert_message_to_dict(m) for m in messages[0]]
+        return message_dicts, params
+
+    def on_chat_model_start(
+        self,
+        serialized: Dict[str, Any],
+        messages: List[List[BaseMessage]],
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        tags: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> Any:
+        self.request_start_time = datetime.datetime.now().timestamp()
+
+        self.messages = messages
+        self.invocation_params = kwargs.get("invocation_params", {})
+        self.name = self.invocation_params.get("_type", "No Type")
+
+    def __init__(self, request_id_func=None, pl_tags=[]):
+        self.request_id_func = request_id_func
+        self.pl_tags = pl_tags
+
+        self.request_start_time = None
+        self.request_end_time = None
+
+        self.prompts = None
+        self.name = None
+        self.invocation_params = None
+
+    def on_llm_start(
```

Comment:
Probably better to accept the run id explicitly and keep this state keyed on the run id